Unfortunately I'm unable to get this to work. I tried:
```rust
impl<UART> SerialExt for UART where UART: Deref<Target = uart1::RegisterBlock> {
    fn constrain(self, baud_rate: Bps, clocks: &Clocks) -> Serial<UART> {
        Serial::<UART> { uart: self } // Just to test
    }
}
```
as well as some other variants, every time it fails with:
```
error[E0119]: conflicting implementations of trait `serial::SerialExt` for type `k210_pac::UARTHS`:
  --> /home/user/maixgo/k210-crates/k210-hal/src/serial.rs:36:1
   |
29 | impl SerialExt for UARTHS {
   | ------------------------- first implementation here
...
36 | impl<UART> SerialExt for UART where UART: Deref<Target = uart1::RegisterBlock> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `k210_pac::UARTHS`
```
I don't understand why it would conflict with UARTHS at all, as it dereferences to completely different kind of register block. I was also unable to reproduce this in the sandbox (https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=739d250c2af00f698fb4fb6ee535807e). It works if everything is defined inline, just not with the external crate. Huh.

--- a/src/clock.rs
+++ b/src/clock.rs
@@ -9,6 +9,7 @@ use crate::time::Hertz;
 #[derive(Clone, Copy)]
 pub struct Clocks {
     cpu: Hertz,
+    apb0: Hertz,
 }
 
 impl Clocks {
@@ -24,12 +25,18 @@ impl Clocks {
         cpu_freq: 390000000
 */
         Self {
-            cpu: Hertz(403_000_000)
+            cpu: Hertz(403_000_000),
+            apb0: Hertz(195_000_000),
         }
     }
 
     /// Returns CPU frequency
     pub fn cpu(&self) -> Hertz {
         Hertz(self.cpu.0)
+    }
+
+    /// Returns APB0 frequency
+    pub fn apb0(&self) -> Hertz {
+        self.apb0
     }
 }

--- a/src/serial.rs
+++ b/src/serial.rs
@@ -7,14 +7,18 @@
 //! * [`UART3`](crate::pac::UART3)
 
 use core::marker::PhantomData;
+use core::ops::Deref;
 
 use embedded_hal::serial;
 use nb;
 use void::Void;
 
-use crate::pac::UARTHS;
+use crate::pac::{UARTHS,uart1,UART1,UART2,UART3};
 use crate::clock::Clocks;
 use crate::time::Bps;
+
+const UART_RECEIVE_FIFO_1: u32 = 0;
+const UART_SEND_FIFO_8: u32 = 3;
 
 /// Extension trait that constrains UART peripherals
 pub trait SerialExt: Sized {
@@ -24,7 +28,28 @@ pub trait SerialExt: Sized {
 
 impl SerialExt for UARTHS {
     fn constrain(self, baud_rate: Bps, clocks: &Clocks) -> Serial<UARTHS> {
-        Serial::new(self, baud_rate, clocks)
+        Serial::<UARTHS>::new(self, baud_rate, clocks)
+    }
+}
+
+/// Trait to be able to generalize over UART1/UART2/UART3
+pub trait UartX: Deref<Target = uart1::RegisterBlock> {
+    /// Return pointer to register block
+    fn ptr() -> *const uart1::RegisterBlock;
+}
+impl UartX for UART1 {
+    fn ptr() -> *const uart1::RegisterBlock { UART1::ptr() }
+}
+impl UartX for UART2 {
+    fn ptr() -> *const uart1::RegisterBlock { UART2::ptr() }
+}
+impl UartX for UART3 {
+    fn ptr() -> *const uart1::RegisterBlock { UART3::ptr() }
+}
+
+impl<UART: UartX> SerialExt for UART {
+    fn constrain(self, baud_rate: Bps, clocks: &Clocks) -> Serial<UART> {
+        Serial::<UART>::new(self, baud_rate, clocks)
     }
 }
 
@@ -141,5 +166,85 @@ impl serial::Write<u8> for Tx<UARTHS> {
         } else {
             Ok(())
         }
+    }
+}
+
+
+impl<UART: UartX> Serial<UART> {
+    /// Configures a UART peripheral to provide serial communication
+    pub fn new(uart: UART, baud_rate: Bps, clocks: &Clocks) -> Self {
+        // Hardcode these for now:
+        let data_width = 8; // 8 data bits
+        let stopbit_val = 0; // 1 stop bit
+        let parity_val = 0; // No parity
+        // Note: need to make sure that UARTx clock is enabled through sysctl before here
+        let divisor = clocks.apb0().0 / baud_rate.0;
+        let dlh = ((divisor >> 12) & 0xff) as u8;
+        let dll = ((divisor >> 4) & 0xff) as u8;
+        let dlf = (divisor & 0xf) as u8;
+        unsafe {
+            // Set Divisor Latch Access Bit (enables DLL DLH) to set baudrate
+            uart.lcr.write(|w| w.bits(1 << 7));
+            uart.dlh_ier.write(|w| w.bits(dlh.into()));
+            uart.rbr_dll_thr.write(|w| w.bits(dll.into()));
+            uart.dlf.write(|w| w.bits(dlf.into()));
+            // Clear Divisor Latch Access Bit after setting baudrate
+            uart.lcr.write(|w| w.bits((data_width - 5) | (stopbit_val << 2) | (parity_val << 3)));
+            // Write IER
+            uart.dlh_ier.write(|w| w.bits(0x80)); /* THRE */
+            // Write FCT
+            uart.fcr_iir.write(|w| w.bits(UART_RECEIVE_FIFO_1 << 6 | UART_SEND_FIFO_8 << 4 | 0x1 << 3 | 0x1));
+        }
+
+        Serial { uart }
+    }
+
+    /// Starts listening for an interrupt event
+    pub fn listen(self) -> Self {
+        self
+    }
+
+    /// Stops listening for an interrupt event
+    pub fn unlisten(self) -> Self {
+        self
+    }
+}
+
+impl<UART: UartX> serial::Read<u8> for Rx<UART> {
+    type Error = Void;
+
+    fn read(&mut self) -> nb::Result<u8, Void> {
+        // NOTE(unsafe) atomic read with no side effects
+        let lsr = unsafe { (*UART::ptr()).lsr.read() };
+
+        if (lsr.bits() & (1<<0)) == 0 { // Data Ready bit
+            Err(::nb::Error::WouldBlock)
+        } else {
+            let rbr = unsafe { (*UART::ptr()).rbr_dll_thr.read() };
+            Ok((rbr.bits() & 0xff) as u8)
+        }
+    }
+}
+
+impl<UART: UartX> serial::Write<u8> for Tx<UART> {
+    type Error = Void;
+
+    fn write(&mut self, byte: u8) -> nb::Result<(), Void> {
+        // NOTE(unsafe) atomic read with no side effects
+        let lsr = unsafe { (*UART::ptr()).lsr.read() };
+
+        if (lsr.bits() & (1<<5)) != 0 { // Transmit Holding Register Empty bit
+            Err(::nb::Error::WouldBlock)
+        } else {
+            unsafe {
+                (*UART::ptr()).rbr_dll_thr.write(|w| w.bits(byte.into()));
+            }
+            Ok(())
+        }
+    }
+
+    fn flush(&mut self) -> nb::Result<(), Void> {
+        // TODO
+        Ok(())
     }
 }

--- a/src/serial.rs
+++ b/src/serial.rs
@@ -6,7 +6,7 @@
 //! * [`UART2`](crate::pac::UART2)
 //! * [`UART3`](crate::pac::UART3)
 
-use core::marker::PhantomData;
+use core::mem;
 use core::ops::Deref;
 
 use embedded_hal::serial;
@@ -65,7 +65,7 @@ pub struct Rx<UART> {
 
 /// Serial transmitter
 pub struct Tx<UART> {
-    _uart: PhantomData<UART>,
+    uart: UART,
 }
 
 impl<UART> Rx<UART> {
@@ -82,7 +82,7 @@ impl<UART> Serial<UART> {
     pub fn split(self) -> (Tx<UART>, Rx<UART>) {
         (
             Tx {
-                _uart: PhantomData
+                uart: unsafe { mem::zeroed() }
             },
             Rx {
                 uart: self.uart
@@ -129,8 +129,7 @@ impl serial::Read<u8> for Rx<UARTHS> {
     type Error = Void;
 
     fn read(&mut self) -> nb::Result<u8, Void> {
-        // NOTE(unsafe) atomic read with no side effects
-        let rxdata = unsafe { (*UARTHS::ptr()).rxdata.read() };
+        let rxdata = self.uart.rxdata.read();
 
         if rxdata.empty().bit_is_set() {
             Err(::nb::Error::WouldBlock)
@@ -144,8 +143,7 @@ impl serial::Write<u8> for Tx<UARTHS> {
     type Error = Void;
 
     fn write(&mut self, byte: u8) -> nb::Result<(), Void> {
-        // NOTE(unsafe) atomic read with no side effects
-        let txdata = unsafe { (*UARTHS::ptr()).txdata.read() };
+        let txdata = self.uart.txdata.read();
 
         if txdata.full().bit_is_set() {
             Err(::nb::Error::WouldBlock)
@@ -158,8 +156,7 @@ impl serial::Write<u8> for Tx<UARTHS> {
     }
 
     fn flush(&mut self) -> nb::Result<(), Void> {
-        // NOTE(unsafe) atomic read with no side effects
-        let txdata = unsafe { (*UARTHS::ptr()).txdata.read() };
+        let txdata = self.uart.txdata.read();
 
         if txdata.full().bit_is_set() {
             Err(nb::Error::WouldBlock)
@@ -214,13 +211,12 @@ impl<UART: UartX> serial::Read<u8> for Rx<UART> {
     type Error = Void;
 
     fn read(&mut self) -> nb::Result<u8, Void> {
-        // NOTE(unsafe) atomic read with no side effects
-        let lsr = unsafe { (*UART::ptr()).lsr.read() };
+        let lsr = self.uart.lsr.read();
 
         if (lsr.bits() & (1<<0)) == 0 { // Data Ready bit
             Err(::nb::Error::WouldBlock)
         } else {
-            let rbr = unsafe { (*UART::ptr()).rbr_dll_thr.read() };
+            let rbr = self.uart.rbr_dll_thr.read();
             Ok((rbr.bits() & 0xff) as u8)
         }
     }
@@ -230,14 +226,13 @@ impl<UART: UartX> serial::Write<u8> for Tx<UART> {
     type Error = Void;
 
     fn write(&mut self, byte: u8) -> nb::Result<(), Void> {
-        // NOTE(unsafe) atomic read with no side effects
-        let lsr = unsafe { (*UART::ptr()).lsr.read() };
+        let lsr = self.uart.lsr.read();
 
         if (lsr.bits() & (1<<5)) != 0 { // Transmit Holding Register Empty bit
             Err(::nb::Error::WouldBlock)
         } else {
             unsafe {
-                (*UART::ptr()).rbr_dll_thr.write(|w| w.bits(byte.into()));
+                self.uart.rbr_dll_thr.write(|w| w.bits(byte.into()));
             }
             Ok(())
         }

--- a/src/serial.rs
+++ b/src/serial.rs
@@ -33,19 +33,10 @@ impl SerialExt for UARTHS {
 }
 
 /// Trait to be able to generalize over UART1/UART2/UART3
-pub trait UartX: Deref<Target = uart1::RegisterBlock> {
-    /// Return pointer to register block
-    fn ptr() -> *const uart1::RegisterBlock;
-}
-impl UartX for UART1 {
-    fn ptr() -> *const uart1::RegisterBlock { UART1::ptr() }
-}
-impl UartX for UART2 {
-    fn ptr() -> *const uart1::RegisterBlock { UART2::ptr() }
-}
-impl UartX for UART3 {
-    fn ptr() -> *const uart1::RegisterBlock { UART3::ptr() }
-}
+pub trait UartX: Deref<Target = uart1::RegisterBlock> { }
+impl UartX for UART1 { }
+impl UartX for UART2 { }
+impl UartX for UART3 { }
 
 impl<UART: UartX> SerialExt for UART {
     fn constrain(self, baud_rate: Bps, clocks: &Clocks) -> Serial<UART> {

